File: src/aizk/graph/writer.py
from collections.abc import Callable
from datetime import datetime
from typing import cast

from loguru import logger
from patos import FlexModel, FrozenModel, sql
from pgvector.sqlalchemy import HALFVEC
from pydantic import UUID5, UUID7, Field, field_validator
from sqlalchemy import (
    Integer,
    Text,
    Uuid,
    column,
    func,
    literal,
    true,
)
from sqlalchemy.dialects.postgresql import Range, insert
from sqlalchemy.sql.selectable import CTE
from sqlmodel import select, tuple_
from sqlmodel.sql.expression import Select

from ..config import settings
from ..extract.models import ConsolidationVerdict, TimedFact
from ..ontology import Ontology
from ..provenance import CaptureContext
from ..store import Entity, Fact, Relation
from ..store.engine import Session
from ..store.models.tables import EntityContent, FactContent
from ..types import Scopes
from .consolidation import Consolidator, FactMatch
from .dedupe import ClaimField
from .grounding import quote_interval
from .ids import entity_id, fact_id
from .naming import normalize_name

# SQLModel synthesizes table-model keyword constructors outside the static signatures.
_entity_content = cast("Callable[..., EntityContent]", Entity.Content)
_fact_content = cast("Callable[..., FactContent]", Fact.Content)


class PreparedEntity(FrozenModel):
    """An extracted entity with its resolved type and precomputed name embedding."""

    name: str
    type: str
    vector: tuple[float, ...]


class FactCandidate(FrozenModel):
    """A new scoped fact whose referenced entities have already resolved."""

    fact: TimedFact
    subject_id: UUID5
    object_id: UUID5 | None
    identity: UUID5


class FactPlan(FrozenModel):
    """A fact candidate with its vector, ranked pool, and rule verdict."""

    candidate: FactCandidate
    vector: tuple[float, ...]
    matches: tuple[FactMatch, ...]
    verdict: ConsolidationVerdict | None


type _Write = tuple[int, FactPlan, ConsolidationVerdict]


class GraphWriter(FlexModel):
    """One graph-write round bound to the exact scope set every write in it shares."""

    session: Session
    created_by: UUID5
    scopes: list[UUID5]
    consolidator: Consolidator
    capture: CaptureContext = Field(default_factory=CaptureContext)
    source_text: str = ""

    @field_validator("scopes", mode="before")
    @classmethod
    def ordered_scopes(cls, scopes: Scopes) -> list[UUID5]:
        """Canonicalize the exact scope set into the sorted list every claim stores."""
        return sorted(scopes)

    def grounding(self, fact: TimedFact) -> dict[str, int]:
        """Char offsets of the fact's supporting quote inside the source chunk, when it
        aligns."""
        interval = quote_interval(fact.quote, self.source_text)
        if interval is None:
            return {}
        return {"quote_start": interval[0], "quote_end": interval[1]}

    async def resolve(self, entity: PreparedEntity) -> UUID5 | None:
        """Resolve one entity through the same set-based path used for a chunk."""
        return (await self.resolve_all([entity])).get(normalize_name(entity.name))

    async def resolve_all(self, entities: list[PreparedEntity]) -> dict[str, UUID5]:
        """Resolve a chunk by normalized identity and return normalized name to content ID.

        Semantic vector proximity is deliberately excluded from identity. Similar concepts
        such as Health, Business, and Research are related graph nodes, not aliases.
        """
        usable: list[tuple[PreparedEntity, UUID5]] = []
        for entity in entities:
            if normalize_name(entity.name):
                usable.append((entity, entity_id(entity.name, entity.type)))
            else:
                logger.warning("entity name {!r} is a path or link, dropping", entity.name)
        if not usable:
            return {}
        contents = {
            node: _entity_content(
                id=node,
                name=entity.name,
                type=entity.type,
                embedding=list(entity.vector),
            )
            for entity, node in usable
        }
        await Entity.Content.mint_all(
            self.session,
            list(contents.values()),
        )
        await Entity.Claim.claim_all(
            self.session,
            list(contents),
            self.created_by,
            frozenset(self.scopes),
        )
        return {normalize_name(entity.name): node for entity, node in usable}

    def candidate(self, fact: TimedFact, resolved: dict[str, UUID5]) -> FactCandidate | None:
        """Build a candidate when both named endpoints resolved to stored entities."""
        subject_id = resolved.get(normalize_name(fact.subject))
        if subject_id is None:
            logger.warning("fact subject {!r} has no resolved entity, skipping", fact.subject)
            return None
        object_id = resolved.get(normalize_name(fact.object_)) if fact.object_ else None
        if fact.object_ and object_id is None:
            logger.warning("fact object {!r} has no resolved entity, skipping", fact.object_)
            return None
        return FactCandidate(
            fact=fact,
            subject_id=subject_id,
            object_id=object_id,
            identity=fact_id(subject_id, fact.predicate, object_id, fact.statement),
        )

    async def new_candidates(
        self, facts: list[TimedFact], resolved: dict[str, UUID5]
    ) -> list[FactCandidate]:
        """The facts not already claimed by this container and whose subject resolved to a
        real entity, the consolidation cascade's first, free tier."""
        candidates = [candidate for fact in facts if (candidate := self.candidate(fact, resolved))]
        # Non-state candidates deduplicate on their exact identity, which already spans
        # subject, predicate, object, and statement, so distinct simultaneous facts all
        # survive. A state relation holds one current value per subject slot and window,
        # so same-slot state candidates inside one batch collapse to the first; letting
        # both through would revise the same current claim twice in one write.
        unique: dict[
            tuple[UUID5, str, datetime | None, datetime | None, str],
            FactCandidate,
        ] = {}
        for candidate in candidates:
            policy = Ontology.current().relation_policies[candidate.fact.predicate]
            state = policy == Relation.Policy.state
            unique.setdefault(
                (
                    candidate.subject_id if state else candidate.identity,
                    candidate.fact.predicate if state else "",
                    candidate.fact.valid_from if state else None,
                    candidate.fact.valid_to if state else None,
                    candidate.fact.kind.perspective_key(self.created_by),
                ),
                candidate,
            )
        candidates = list(unique.values())
        if not candidates:
            return []
        keys = [
            (candidate.identity, candidate.fact.kind.perspective_key(self.created_by))
            for candidate in candidates
        ]
        claimed = set(
            await self.session.exec(
                select(Fact.Claim.content_id, Fact.Claim.perspective_key).where(
                    Fact.Claim.scopes == self.scopes,
                    tuple_(Fact.Claim.content_id, Fact.Claim.perspective_key).in_(keys),
                )
            )
        )
        return [
            candidate
            for candidate in candidates
            if (
                candidate.identity,
                candidate.fact.kind.perspective_key(self.created_by),
            )
            not in claimed
        ]

    async def plan_facts(
        self,
        candidates: list[FactCandidate],
        vectors: list[list[float]],
    ) -> list[FactPlan]:
        """Rank the narrow top fact matches in PostgreSQL and apply deterministic verdicts."""
        if not candidates:
            return []
        matches = await self._fact_matches(self._fact_inputs(candidates, vectors), len(candidates))
        return [
            FactPlan(
                candidate=candidate,
                vector=tuple(vector),
                matches=tuple(ranked),
                verdict=self.consolidator.decide(
                    Ontology.current().relation_policies[candidate.fact.predicate],
                    candidate.object_id,
                    ranked,
                ),
            )
            for candidate, vector, ranked in zip(candidates, vectors, matches, strict=True)
        ]

    def _fact_inputs(self, candidates: list[FactCandidate], vectors: list[list[float]]) -> CTE:
        """Render fact candidates as one typed input relation."""
        return sql.relation(
            "fact_input",
            (
                column("ordinal", Integer),
                column("subject_id", Uuid),
                column("predicate", Text),
                column("perspective_key", Text),
                column("embedding", HALFVEC(settings.embed_dim)),
            ),
            [
                (
                    ordinal,
                    candidate.subject_id,
                    candidate.fact.predicate,
                    candidate.fact.kind.perspective_key(self.created_by),
                    vector,
                )
                for ordinal, (candidate, vector) in enumerate(
                    zip(candidates, vectors, strict=True)
                )
            ],
        )

    async def _fact_matches(self, inputs: CTE, count: int) -> list[list[FactMatch]]:
        """Return each input's nearest live facts from one lateral database query."""
        distance = Fact.Content.embedding @ inputs.c.embedding.cast(HALFVEC(settings.embed_dim))
        ranked = (
            select(
                Fact.Claim.id,
                Fact.Content.object_id,
                Fact.Content.statement,
                distance.label("distance"),
            )
            .select_from(Fact.Content.__table__.join(Fact.Claim.__table__))
            .where(
                Fact.Content.subject_id == inputs.c.subject_id,
                Fact.Content.predicate == inputs.c.predicate,
                Fact.Claim.scopes == self.scopes,
                Fact.Claim.perspective_key == inputs.c.perspective_key,
                Fact.Claim.is_current,
            )
            .order_by(distance)
            .limit(settings.similar_facts)
            .lateral("ranked_fact")
        )
        matches: list[list[FactMatch]] = [[] for _ in range(count)]
        rows = await self.session.exec(
            # `add_columns` erases the sqlmodel `Select` subtype statically while the
            # runtime object stays one, so the cast restores what `exec` requires.
            cast(
                "Select[tuple[int, UUID7 | None, UUID5 | None, str | None, float | None]]",
                select(
                    inputs.c.ordinal,
                    ranked.c.id,
                    ranked.c.object_id,
                    ranked.c.statement,
                )
                .add_columns(ranked.c.distance)
                .select_from(inputs.outerjoin(ranked, true()))
                .order_by(inputs.c.ordinal, ranked.c.distance),
            )
        )
        for ordinal, claim_id, object_id, statement, match_distance in rows:
            if claim_id is not None:
                matches[ordinal].append(
                    FactMatch(
                        id=claim_id,
                        object_id=object_id,
                        statement=cast(str, statement),
                        distance=cast(float, match_distance),
                    )
                )
        return matches

    def borderline(self, plans: list[FactPlan]) -> list[tuple[TimedFact, list[FactMatch]]]:
        """Return only plans whose similarity needs one batched LLM decision."""
        return [
            (plan.candidate.fact, list(plan.matches)) for plan in plans if plan.verdict is None
        ]

    async def resolve_ambiguous(self, plans: list[FactPlan]) -> list[ConsolidationVerdict]:
        """Resolve the plans left undecided by deterministic similarity rules."""
        return await self.consolidator.resolve(self.borderline(plans))

    async def lock_plans(self, plans: list[FactPlan]) -> None:
        """Serialize each scope, subject, and perspective slot for final revalidation."""
        slots = sorted(
            {
                (
                    plan.candidate.subject_id,
                    plan.candidate.fact.predicate,
                    plan.candidate.fact.kind.perspective_key(self.created_by),
                )
                for plan in plans
            },
            key=lambda slot: (str(slot[0]), slot[1], slot[2]),
        )
        if not slots:
            return
        inputs = sql.relation(
            "fact_lock",
            (
                column("subject_id", Uuid),
                column("predicate", Text),
                column("perspective_key", Text),
            ),
            slots,
        )
        key = func.hashtextextended(
            inputs.c.subject_id.cast(Text)
            + literal("|")
            + inputs.c.predicate
            + literal("|")
            + inputs.c.perspective_key
            + literal("|")
            + literal(",".join(str(scope) for scope in self.scopes)),
            0,
        )
        await self.session.exec(
            select(func.pg_advisory_xact_lock(key)).select_from(inputs).order_by(key)
        )

    async def apply_plans(
        self,
        plans: list[FactPlan],
        resolved: list[ConsolidationVerdict],
        source_chunk_id: UUID7,
    ) -> None:
        """Apply already-decided plans inside one short write transaction."""
        verdicts = merged_verdicts([plan.verdict for plan in plans], resolved)
        writes = self._writes(plans, verdicts)
        valid_ends = await Fact.Claim.revise(
            self.session,
            self._revisions(writes),
        )
        rows = self._write_rows(writes, valid_ends, source_chunk_id)
        contents = [content for content, _ in rows]
        claims = [claim for _, claim in rows]
        await Fact.Content.mint_all(self.session, contents)
        if claims:
            await self.session.exec(
                insert(Fact.Claim)
                .values(claims)
                .on_conflict_do_nothing(
                    index_elements=[
                        Fact.Claim.content_id,
                        Fact.Claim.scopes,
                        Fact.Claim.perspective_key,
                    ],
                    index_where=Fact.Claim.recorded.f.upper_inf(result=bool),
                )
            )

    @staticmethod
    def _writes(plans: list[FactPlan], verdicts: list[ConsolidationVerdict]) -> list[_Write]:
        """Pair every non-noop decision with its stable input ordinal."""
        return [
            (ordinal, plan, verdict)
            for ordinal, (plan, verdict) in enumerate(zip(plans, verdicts, strict=True))
            if verdict.action != "NOOP"
        ]

    @staticmethod
    def _revisions(
        writes: list[_Write],
    ) -> list[tuple[int, UUID7, datetime | None, datetime | None]]:
        """Close every occupied state slot and only the chosen claim for other updates."""
        revisions: list[tuple[int, UUID7, datetime | None, datetime | None]] = []
        for ordinal, plan, verdict in writes:
            if verdict.action != "UPDATE" or verdict.supersedes is None:
                continue
            policy = Ontology.current().relation_policies[plan.candidate.fact.predicate]
            claims = (
                [match.id for match in plan.matches]
                if policy == Relation.Policy.state
                else [verdict.supersedes]
            )
            revisions.extend(
                (
                    ordinal,
                    claim_id,
                    plan.candidate.fact.valid_from,
                    plan.candidate.fact.valid_to,
                )
                for claim_id in claims
            )
        return revisions

    def _write_rows(
        self,
        writes: list[_Write],
        valid_ends: dict[int, datetime | None],
        source_chunk_id: UUID7,
    ) -> list[tuple[FactContent, dict[str, ClaimField]]]:
        """Render decided writes using the temporal ends returned by PostgreSQL."""
        return [
            self._write_row(
                plan,
                source_chunk_id,
                valid_ends[ordinal]
                if verdict.action == "UPDATE" and verdict.supersedes is not None
                else plan.candidate.fact.valid_to,
            )
            for ordinal, plan, verdict in writes
        ]

    def _write_row(
        self,
        plan: FactPlan,
        source_chunk_id: UUID7,
        valid_to: datetime | None,
    ) -> tuple[FactContent, dict[str, ClaimField]]:
        """Render one decided fact as immutable content and its scoped temporal claim."""
        candidate = plan.candidate
        fact = candidate.fact
        return (
            _fact_content(
                id=candidate.identity,
                subject_id=candidate.subject_id,
                object_id=candidate.object_id,
                predicate=fact.predicate,
                statement=fact.statement,
                embedding=list(plan.vector),
            ),
            {
                "content_id": candidate.identity,
                "created_by": self.created_by,
                "scopes": self.scopes,
                "valid": Range(fact.valid_from, valid_to) if fact.valid_from or valid_to else None,
                "source_chunk_id": source_chunk_id,
                "attributes": self.capture.claim_attributes(fact.kind, self.created_by)
                | self.grounding(fact),
                "perspective_key": fact.kind.perspective_key(self.created_by),
            },
        )


def merged_verdicts(
    verdicts: list[ConsolidationVerdict | None], resolved: list[ConsolidationVerdict]
) -> list[ConsolidationVerdict]:
    """Fill each null, genuinely-ambiguous slot with the batched LLM's own verdict, in order."""
    pending = iter(resolved)
    return [verdict if verdict is not None else next(pending) for verdict in verdicts]

File: archive/ts/src/lib/server/graph/writer.ts
// One graph-write round bound to the exact scope set every write in it shares, ported from
// graph/writer.py. Entity resolution is one VALUES-CTE read with a lateral nearest-entity
// probe plus bulk content and claim writes; fact planning ranks the narrow top matches in
// PostgreSQL through a lateral per candidate; lock_plans serializes each scope, subject and
// perspective slot through hashtextextended advisory locks; and apply_plans writes decided
// plans with the supersede valid-range bookkeeping and live-index on-conflict claim
// inserts. Content inserts land as ON CONFLICT DO NOTHING, the set-level twin of the
// Python mint_all savepoint dance with the same end state.
import { sql, type SQL } from 'drizzle-orm';
import { v7 as uuid7 } from 'uuid';

import { quoteInterval } from '../extract/grounding';
import {
	claimAttributes,
	perspectiveKey,
	type CaptureContext,
	type ConsolidationVerdict,
	type JsonValue,
	type TimedFact
} from '../extract/models';
import { vectorLiteral } from '../serving';
import { settings } from '../settings';
import type { Tx } from '../db';
import { entityClaim, entityContent, factClaim, factContent } from '../db/schema';
import { decideByRule, type FactMatch } from './consolidation';
import { entityId, factId } from './ids';
import { normalizeName } from './naming';
import { uuidArray } from './system';

const n = (value: number): SQL => sql.raw(value.toFixed(0));

// A '[)' tstzrange literal with optional bounds, the asyncpg Range default rendering.
const rangeLiteral = (lower: Date | null, upper: Date | null): string =>
	`[${lower === null ? '' : `"${lower.toISOString()}"`},${upper === null ? '' : `"${upper.toISOString()}"`})`;

// An extracted entity with its resolved type and precomputed name embedding.
export interface PreparedEntity {
	name: string;
	type: string;
	vector: number[];
}

// A new scoped fact whose referenced entities have already resolved.
export interface FactCandidate {
	fact: TimedFact;
	subjectId: string;
	objectId: string | null;
	identity: string;
}

// A fact candidate with its vector, ranked pool, and rule verdict.
export interface FactPlan {
	candidate: FactCandidate;
	vector: number[];
	matches: FactMatch[];
	verdict: ConsolidationVerdict | null;
}

// Fill each null, genuinely-ambiguous slot with the batched LLM's own verdict, in order.
export const mergedVerdicts = (
	verdicts: (ConsolidationVerdict | null)[],
	resolved: ConsolidationVerdict[]
): (ConsolidationVerdict | null)[] => {
	let next = 0;
	return verdicts.map((verdict) => verdict ?? resolved[next++]);
};

export class GraphWriter {
	readonly createdBy: string;
	readonly scopes: string[];
	readonly capture: CaptureContext;
	readonly sourceText: string;

	constructor(createdBy: string, scopes: string[], capture: CaptureContext, sourceText = '') {
		this.createdBy = createdBy;
		this.scopes = [...new Set(scopes)].sort();
		this.capture = capture;
		this.sourceText = sourceText;
	}

	// Char offsets of the fact's supporting quote inside the source chunk, when it aligns.
	grounding(fact: TimedFact): Record<string, JsonValue> {
		const interval = quoteInterval(fact.quote, this.sourceText);
		if (interval === null) return {};
		return { quote_start: interval[0], quote_end: interval[1] };
	}

	// Resolve a chunk's entities with one claim read and bulk content and claim writes.
	async resolveAll(tx: Tx, entities: PreparedEntity[]): Promise<Map<string, string>> {
		const usable: [PreparedEntity, string][] = [];
		for (const entity of entities) {
			if (normalizeName(entity.name)) usable.push([entity, entityId(entity.name, entity.type)]);
			else console.warn(`entity name ${JSON.stringify(entity.name)} is a path or link, dropping`);
		}
		if (!usable.length) return new Map();
		const inputs = usable.map(
			([entity, node], ordinal) =>
				sql`(${n(ordinal)}::integer, ${node}::uuid, ${entity.type}::text,
					${vectorLiteral(entity.vector)}::halfvec(${n(settings.embedDim)}))`
		);
		const rows = (await tx.execute(sql`
			with entity_input (ordinal, id, type, embedding) as (values ${sql.join(inputs, sql`, `)})
			select ei.ordinal as ordinal,
				coalesce(cl.content_id, nearest.id, ei.id) as resolved,
				(cl.content_id is null and nearest.id is null) as is_new
			from entity_input ei
			left join entity_claim cl
				on cl.content_id = ei.id and cl.scopes = ${uuidArray(this.scopes)}::uuid[]
			left join lateral (
				select ec.id from entity_content ec
				where ec.type = ei.type
					and (ec.embedding <=> ei.embedding) <= ${1.0 - settings.entityResolutionThreshold}
				order by ec.embedding <=> ei.embedding
				limit 1
			) nearest on cl.content_id is null
			order by ei.ordinal
		`)) as unknown as { ordinal: number; resolved: string; is_new: boolean }[];
		const resolved = new Map<string, string>();
		const newContents = new Map<string, PreparedEntity>();
		for (const row of rows) {
			const [entity, node] = usable[row.ordinal];
			resolved.set(entity.name, row.resolved);
			if (row.is_new) newContents.set(node, entity);
		}
		if (newContents.size)
			await tx
				.insert(entityContent)
				.values(
					[...newContents.entries()].map(([node, entity]) => ({
						id: node,
						name: entity.name,
						type: entity.type,
						embedding: vectorLiteral(entity.vector)
					}))
				)
				.onConflictDoNothing();
		const claimIds = [...new Set(resolved.values())];
		await tx
			.insert(entityClaim)
			.values(
				claimIds.map((contentId) => ({
					id: uuid7(),
					contentId,
					createdBy: this.createdBy,
					scopes: this.scopes
				}))
			)
			.onConflictDoNothing({ target: [entityClaim.contentId, entityClaim.scopes] });
		return resolved;
	}

	// Build a candidate when its subject resolved to a stored entity.
	candidate(fact: TimedFact, resolved: Map<string, string>): FactCandidate | null {
		const subjectId = resolved.get(fact.subject);
		if (subjectId === undefined) {
			console.warn(`fact subject ${JSON.stringify(fact.subject)} has no resolved entity, skipping`);
			return null;
		}
		const objectId = fact.object ? (resolved.get(fact.object) ?? null) : null;
		return {
			fact,
			subjectId,
			objectId,
			identity: factId(fact.subject, fact.predicate, fact.object, fact.statement)
		};
	}

	// The facts not already claimed by this container and whose subject resolved to a real
	// entity, the consolidation cascade's first, free tier. Python's ORM read runs under the
	// live temporal gate, so only current claims block a candidate here.
	async newCandidates(
		tx: Tx,
		facts: TimedFact[],
		resolved: Map<string, string>
	): Promise<FactCandidate[]> {
		const candidates = facts
			.map((fact) => this.candidate(fact, resolved))
			.filter((candidate): candidate is FactCandidate => candidate !== null);
		if (!candidates.length) return [];
		const keys = candidates.map(
			(candidate) =>
				sql`(${candidate.identity}::uuid, ${perspectiveKey(candidate.fact.kind, this.createdBy)}::varchar)`
		);
		const rows = (await tx.execute(sql`
			select content_id, perspective_key from fact_claim
			where scopes = ${uuidArray(this.scopes)}::uuid[]
				and (content_id, perspective_key) in (values ${sql.join(keys, sql`, `)})
				and upper_inf(recorded) and (valid is null or valid @> now())
		`)) as unknown as { content_id: string; perspective_key: string }[];
		const claimed = new Set(rows.map((row) => `${row.content_id}|${row.perspective_key}`));
		return candidates.filter(
			(candidate) =>
				!claimed.has(
					`${candidate.identity}|${perspectiveKey(candidate.fact.kind, this.createdBy)}`
				)
		);
	}

	// Rank the narrow top fact matches in PostgreSQL and apply deterministic verdicts.
	async planFacts(tx: Tx, candidates: FactCandidate[], vectors: number[][]): Promise<FactPlan[]> {
		if (!candidates.length) return [];
		const inputs = candidates.map(
			(candidate, ordinal) =>
				sql`(${n(ordinal)}::integer, ${candidate.subjectId}::uuid,
					${perspectiveKey(candidate.fact.kind, this.createdBy)}::text,
					${vectorLiteral(vectors[ordinal])}::halfvec(${n(settings.embedDim)}))`
		);
		const rows = (await tx.execute(sql`
			with fact_input (ordinal, subject_id, perspective_key, embedding) as (
				values ${sql.join(inputs, sql`, `)}
			)
			select fi.ordinal as ordinal, ranked.id as id, ranked.predicate as predicate,
				ranked.object_id as object_id, ranked.statement as statement, ranked.distance as distance
			from fact_input fi
			left join lateral (
				select cl.id, fc.predicate, fc.object_id, fc.statement,
					(fc.embedding <=> fi.embedding) as distance
				from fact_content fc
				join fact_claim cl on cl.content_id = fc.id
				where fc.subject_id = fi.subject_id
					and cl.scopes = ${uuidArray(this.scopes)}::uuid[]
					and cl.perspective_key = fi.perspective_key
					and upper_inf(cl.recorded) and (cl.valid is null or cl.valid @> now())
				order by distance
				limit ${n(settings.similarFacts)}
			) ranked on true
			order by fi.ordinal, ranked.distance
		`)) as unknown as {
			ordinal: number;
			id: string | null;
			predicate: string | null;
			object_id: string | null;
			statement: string | null;
			distance: number | null;
		}[];
		const matches: FactMatch[][] = candidates.map(() => []);
		for (const row of rows) {
			if (row.id === null) continue;
			matches[row.ordinal].push({
				id: row.id,
				predicate: row.predicate as string,
				objectId: row.object_id,
				statement: row.statement as string,
				distance: Number(row.distance)
			});
		}
		return candidates.map((candidate, ordinal) => ({
			candidate,
			vector: vectors[ordinal],
			matches: matches[ordinal],
			verdict: decideByRule(candidate.fact.predicate, candidate.objectId, matches[ordinal])
		}));
	}

	// Return only plans whose similarity needs one batched LLM decision.
	borderline(plans: FactPlan[]): [TimedFact, FactMatch[]][] {
		return plans
			.filter((plan) => plan.verdict === null)
			.map((plan) => [plan.candidate.fact, plan.matches]);
	}

	// Serialize each scope, subject, and perspective slot for final revalidation.
	async lockPlans(tx: Tx, plans: FactPlan[]): Promise<void> {
		const slots = [
			...new Map(
				plans.map((plan) => {
					const key = perspectiveKey(plan.candidate.fact.kind, this.createdBy);
					return [`${plan.candidate.subjectId}|${key}`, [plan.candidate.subjectId, key]];
				})
			).values()
		].sort(([a, aKey], [b, bKey]) => (a < b ? -1 : a > b ? 1 : aKey < bKey ? -1 : 1));
		if (!slots.length) return;
		const inputs = slots.map(
			([subjectId, key]) => sql`(${subjectId}::uuid, ${key}::text)`
		);
		const joined = this.scopes.join(',');
		const key = (alias: SQL): SQL =>
			sql`hashtextextended(${alias}.subject_id::text || '|' || ${alias}.perspective_key || '|' || ${joined}, 0)`;
		await tx.execute(sql`
			with fact_lock (subject_id, perspective_key) as (values ${sql.join(inputs, sql`, `)})
			select pg_advisory_xact_lock(${key(sql.raw('fl'))})
			from fact_lock fl
			order by ${key(sql.raw('fl'))}
		`);
	}

	// Apply already-decided plans inside one short write transaction.
	async applyPlans(
		tx: Tx,
		plans: FactPlan[],
		resolved: ConsolidationVerdict[],
		sourceChunkId: string
	): Promise<void> {
		const verdicts = mergedVerdicts(
			plans.map((plan) => plan.verdict),
			resolved
		);
		const superseded = verdicts
			.filter(
				(verdict): verdict is ConsolidationVerdict =>
					verdict !== null && verdict.action === 'UPDATE' && verdict.supersedes !== null
			)
			.map((verdict) => verdict.supersedes as string);
		const retired = new Map<string, { validLower: Date | null }>();
		if (superseded.length) {
			const rows = (await tx.execute(sql`
				select id, lower(valid) as valid_lower from fact_claim
				where id in (select (value)::uuid from jsonb_array_elements_text(${JSON.stringify(superseded)}::jsonb))
			`)) as unknown as { id: string; valid_lower: Date | null }[];
			for (const row of rows)
				retired.set(row.id, { validLower: row.valid_lower === null ? null : new Date(row.valid_lower) });
		}
		const contents: (typeof factContent.$inferInsert)[] = [];
		const claims: (typeof factClaim.$inferInsert)[] = [];
		for (let index = 0; index < plans.length; index += 1) {
			const plan = plans[index];
			const verdict = verdicts[index];
			if (verdict === null) throw new Error('unresolved consolidation verdict');
			if (verdict.action === 'NOOP') continue;
			const { candidate } = plan;
			const { fact } = candidate;
			const now = new Date();
			let validTo = fact.validTo;
			if (verdict.action === 'UPDATE' && verdict.supersedes !== null) {
				const previous = retired.get(verdict.supersedes);
				if (previous === undefined) throw new Error(`superseded claim ${verdict.supersedes} not found`);
				const lower = previous.validLower;
				if (fact.validFrom !== null && lower !== null && fact.validFrom < lower) {
					if (validTo === null || lower < validTo) validTo = lower;
				} else {
					let closing = fact.validFrom ?? now;
					if (lower !== null && closing < lower) closing = lower;
					await tx.execute(sql`
						update fact_claim
						set valid = ${rangeLiteral(lower, closing)}::tstzrange,
							recorded = tstzrange(lower(recorded), ${now.toISOString()}::timestamptz)
						where id = ${verdict.supersedes}::uuid
					`);
				}
			}
			contents.push({
				id: candidate.identity,
				subjectId: candidate.subjectId,
				objectId: candidate.objectId,
				predicate: fact.predicate,
				statement: fact.statement,
				embedding: vectorLiteral(plan.vector)
			});
			claims.push({
				id: uuid7(),
				contentId: candidate.identity,
				createdBy: this.createdBy,
				scopes: this.scopes,
				valid:
					fact.validFrom !== null || validTo !== null
						? rangeLiteral(fact.validFrom, validTo)
						: null,
				sourceChunkId,
				attributes: {
					...claimAttributes(this.capture, fact.kind, this.createdBy),
					...this.grounding(fact)
				},
				perspectiveKey: perspectiveKey(fact.kind, this.createdBy)
			});
		}
		if (contents.length) await tx.insert(factContent).values(contents).onConflictDoNothing();
		if (claims.length)
			await tx
				.insert(factClaim)
				.values(claims)
				.onConflictDoNothing({
					target: [factClaim.contentId, factClaim.scopes, factClaim.perspectiveKey],
					where: sql`upper_inf(recorded)`
				});
	}
}
